README.md
# Jwt_c

[![Build Status](https://travis-ci.org/issuu/ocaml-libjwt.svg?branch=master)](https://travis-ci.org/issuu/ocaml-libjwt)

OCaml bindings for [libjwt](https://github.com/benmcollins/libjwt)

## License

Apache 2.0, see [LICENSE](LICENSE).


CHANGES.md
0.4
===
* Fully switch from JBuilder to dune
* Generate .opam file using dune (requires dune >= 1.11)

0.3
===
* All functions that previously may have raised
  an exception now return a result instead.

0.2
===
* Add depexts to OPAM file

0.1
===
* Initial release


test/test_jwt_c.ml
module Jwt = Jwt_c

type error = [ `Jwt_error of int ] [@@deriving eq, show]

let error = Alcotest.testable pp_error equal_error

(* the key is just something I generated with `openssl genrsa -out private.pem 2048` *)
let private_key = "-----BEGIN RSA PRIVATE KEY-----
MIIEpQIBAAKCAQEA2EobPLULbXZNI9GBoXvX2X4Yplh4KIhDBWd33ll4id+Wsw6/
vIvOKy2rB74zcTgmRi6ZVNArgX5jdC6+nJHer33DQl/CNls24n7H5b/gC9O253ro
JjYQU9FiwMH4nsneAJimXizNc00E5pqhvMbCu1/ZoKb6hLBrcLF1KA3n38GTcmaq
r9yI5TM65B2pKelNAbBaIB2oBWAKjPfvDLkOhyNrgAKV677LD3fsU+lC+ea9ylt/
KPoicOAQFZer1kqYcI+IeouG2ePJcPEH3xpH0mFtaTzmJd5Hzr84kmM2NZRN+oOl
OYOKFluSTWQ1aojPQBmhceQFnfLAKoOW4qS6VQIDAQABAoIBAQCh8NJ14KP+wD2/
nMSvUNdrS6NPIRxOVL/BkxfHzjuXF2ZxCJD8rbyezaGpuXowwja+A3PgccCxQx1Z
xwoGlp0hzkrdLm9uXVs5uG0ZE1G/6TOgG4En4wDUkQichF8PHNvwnFqsVmU5eCg8
NPj2K+dvfbOnOn6FzMWU6flrFQZQm/HCB0igqyd85kfb9/Cj9iT2Jf9hRA1fhhBE
q+VyHISSD9QOO3E3S8mrHMc6otZdIwT8MWMh0YaJsYSV4/cPe9nVODV9PLPkVgz4
COZcGYfBDElvWo5IPZRcD47rdJx9j2eqnE4dhAOsFSH2qgsQKU566ZAX14DUUCZ6
Pk+6uhRdAoGBAPCyoK94iZ5A9+wc9MdpgyRKlq7RA2zt0GR+47+Liv614+zJAnc1
qIdEqIcDPdMILQbWBIMwtwxYiSQ1ILuL6IfMIsXz3Jpd+BGnV3EtJFg0/TshPu/U
hQsgOjShajemzP8oT3AGcqWmCa6xH27wbPgT9gIa7XkS64kRqBOLZCATAoGBAOYK
O3J0RDgUTWWWBT2INNIn9S1/B7TmHbER3ESCP+f0k5GRvK5LLAqWNrrdGWN5s/xQ
tRWMYOfYDbkyAsolS3vKpo/m3upkTD/G8J0faA/P30wvq3XOEKstbqeiYlAofeVz
TI1sHAutJx0OtOdQB9L/YDbb7JCqY3dBZqeG2Bj3AoGAWyF2fxknGZjFhUrtTnbf
ZSUsaeHO5zYfGX2FYydFJ+zb7/GnElVpilVvbTbH+Jd23Mi5CqauF2mJ+wB2dSui
jY+3drU+x99eJejyzXHm+dKOMg4DUzBmcvDvuK+IlKt9n/m2Idb/H0J/FfoPyaQT
zdVY5jElyhpkvQ/HUCcNcKsCgYEAqUfQVTpP1UDDb8UFGDG6RQhYeOFo52sLYHk3
MUb0BhpJ2a54PX5d63B81+fKPhSaKUuu5iuGSDYKa0TtHppxQKhxB6YqViv2nFwm
RkmNR01+Ec9mimtYgs8NBdkOJdSWSJofNFbhEIqcJNrkru+Kwm9g+x29qPtp9KEx
DIjDTyMCgYEAkdv4S187YxFAXGgcHAKVBP1awardfyspc1sUBa/ABGQl6JRLlI/W
xNCvRhfRxa5mCCLYhCRXIm74JEQC7craX+K0ln83YZ7870HuwNEanuqkIRQhMP1g
qA2xRVxCNmkZ/Ju6i1dXwmVomoECSAYRLqHbRHScSfqqFr2rAec9BQg=
-----END RSA PRIVATE KEY-----"

let public_key = "-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA2EobPLULbXZNI9GBoXvX
2X4Yplh4KIhDBWd33ll4id+Wsw6/vIvOKy2rB74zcTgmRi6ZVNArgX5jdC6+nJHe
r33DQl/CNls24n7H5b/gC9O253roJjYQU9FiwMH4nsneAJimXizNc00E5pqhvMbC
u1/ZoKb6hLBrcLF1KA3n38GTcmaqr9yI5TM65B2pKelNAbBaIB2oBWAKjPfvDLkO
hyNrgAKV677LD3fsU+lC+ea9ylt/KPoicOAQFZer1kqYcI+IeouG2ePJcPEH3xpH
0mFtaTzmJd5Hzr84kmM2NZRN+oOlOYOKFluSTWQ1aojPQBmhceQFnfLAKoOW4qS6
VQIDAQAB
-----END PUBLIC KEY-----"

let get_set_grant_string () =
  match Jwt.create () with
  | Ok jwt ->
    Alcotest.(check (option string)) "non-existing" None (Jwt.get_grant jwt "non-existing");
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grant jwt "abc" "efg"));
    Alcotest.(check (option string)) "existing" (Some "efg") (Jwt.get_grant jwt "abc");
    Gc.full_major ()
  | Error _ ->
    Alcotest.fail "Unable to create JWT"

let get_set_grant_int () =
  match Jwt.create () with
  | Ok jwt ->
    Alcotest.(check (option int)) "non-existing" None (Jwt.get_grant_int jwt "non-existing");
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grant_int jwt "n" 42));
    Alcotest.(check (option int)) "existing" (Some 42) (Jwt.get_grant_int jwt "n");
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grant_int jwt "zero" 0));
    Alcotest.(check (option int)) "zero" (Some 0) (Jwt.get_grant_int jwt "zero");
    Gc.full_major ()
  | Error _ ->
    Alcotest.fail "Unable to create JWT"

let get_set_grant_bool () =
  match Jwt.create () with
  | Ok jwt ->
    Alcotest.(check (option bool)) "non-existing" None (Jwt.get_grant_bool jwt "non-existing");
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grant_bool jwt "true" true));
    Alcotest.(check (option bool)) "true" (Some true) (Jwt.get_grant_bool jwt "true");
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grant_bool jwt "false" false));
    Alcotest.(check (option bool)) "false" (Some false) (Jwt.get_grant_bool jwt "false");
    Gc.full_major ()
  | Error _ ->
    Alcotest.fail "Unable to create JWT"

let get_json () =
  match Jwt.create () with
  | Ok jwt ->
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grant jwt "abc" "efg"));
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grant_int jwt "n" 42));
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grant_bool jwt "b" true));
    Alcotest.(check (option string)) "json" (Some {json|{"abc":"efg","b":true,"n":42}|json}) (Jwt.get_grants_json jwt);
    Alcotest.(check (option string)) "with key" (Some {json|"efg"|json}) (Jwt.get_grants_json ~key:"abc" jwt);
    Alcotest.(check (option string)) "non-existing" None (Jwt.get_grants_json ~key:"non-existing" jwt);
    Gc.full_major ()
  | Error _ ->
    Alcotest.fail "Unable to create JWT"

let set_json () =
  match Jwt.create () with
  | Ok jwt ->
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grants_json jwt {json|{"abc":"efg","b":true,"n":42}|json}));
    Alcotest.(check (option string)) "json" (Some {json|{"abc":"efg","b":true,"n":42}|json}) (Jwt.get_grants_json jwt);
    Alcotest.(check (option string)) "get grant" (Some "efg") (Jwt.get_grant jwt "abc");
    Gc.full_major ()
  | Error _ ->
    Alcotest.fail "Unable to create JWT"

let delete_grant () =
  match Jwt.create () with
  | Ok jwt ->
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grants_json jwt {json|{"abc":"efg","b":true,"n":42}|json}));
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.del_grant jwt "abc"));
    Alcotest.(check (option string)) "json" (Some {json|{"b":true,"n":42}|json}) (Jwt.get_grants_json jwt);
    Gc.full_major ()
  | Error _ ->
    Alcotest.fail "Unable to create JWT"

let delete_all_grants () =
  match Jwt.create () with
  | Ok jwt ->
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grants_json jwt {json|{"abc":"efg","b":true,"n":42}|json}));
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.del_grants jwt));
    Alcotest.(check (option string)) "json" (Some "{}") (Jwt.get_grants_json jwt);
    Gc.full_major ()
  | Error _ ->
    Alcotest.fail "Unable to create JWT"

let decode_encode () =
  match Jwt.create () with
  | Ok jwt ->
    let now = Unix.time () |> int_of_float in
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.set_alg jwt RS256 ~key:private_key));
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grant jwt "iss" "https://www.googleapis.com/robot/v1/metadata/x509/recommendation-tracking%40recommendation-tracking.iam.gserviceaccount.com"));
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grant jwt "scope" "https://www.googleapis.com/auth/prediction"));
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grant jwt "aud" "https://www.googleapis.com/oauth2/v4/token"));
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grant_int jwt "exp" (now + 3600)));
    Alcotest.(check (result unit error) "grant-added" (Ok ()) (Jwt.add_grant_int jwt "iat" now));
    begin match Jwt.encode jwt with
      | Ok jwt_encoded ->
        begin match Jwt.decode jwt_encoded ~key:public_key with
        | Ok jwt_decoded ->
          Alcotest.(check (option string)) "json" (Jwt.get_grants_json jwt) (Jwt.get_grants_json jwt_decoded);
          Gc.full_major ()
        | Error _ ->
          Alcotest.fail "Unable to decode JWT"
        end
      | Error _ ->
        Alcotest.fail "Unable to encode JWT"
    end
  | Error _ ->
    Alcotest.fail "Unable to create JWT"

let () =
  Alcotest.run "Jwt_C" [
    "tests", [
      "get/set grant (string)", `Quick, get_set_grant_string;
      "get/set grant (int)", `Quick, get_set_grant_int;
      "get/set grant (bool)", `Quick, get_set_grant_bool;
      "get json", `Quick, get_json;
      "set json", `Quick, set_json;
      "delete grant", `Quick, delete_grant;
      "delete all grants", `Quick, delete_all_grants;
      "decode / encode", `Quick, decode_encode;
    ]
  ]


test/test_libjwt.ml
open Lwt.Infix
open Cohttp
open Cohttp_lwt_unix

(* the key is just something I generated with `openssl genrsa -out private.pem 2048` *)
let private_key = "-----BEGIN RSA PRIVATE KEY-----
MIIEpQIBAAKCAQEA2EobPLULbXZNI9GBoXvX2X4Yplh4KIhDBWd33ll4id+Wsw6/
vIvOKy2rB74zcTgmRi6ZVNArgX5jdC6+nJHer33DQl/CNls24n7H5b/gC9O253ro
JjYQU9FiwMH4nsneAJimXizNc00E5pqhvMbCu1/ZoKb6hLBrcLF1KA3n38GTcmaq
r9yI5TM65B2pKelNAbBaIB2oBWAKjPfvDLkOhyNrgAKV677LD3fsU+lC+ea9ylt/
KPoicOAQFZer1kqYcI+IeouG2ePJcPEH3xpH0mFtaTzmJd5Hzr84kmM2NZRN+oOl
OYOKFluSTWQ1aojPQBmhceQFnfLAKoOW4qS6VQIDAQABAoIBAQCh8NJ14KP+wD2/
nMSvUNdrS6NPIRxOVL/BkxfHzjuXF2ZxCJD8rbyezaGpuXowwja+A3PgccCxQx1Z
xwoGlp0hzkrdLm9uXVs5uG0ZE1G/6TOgG4En4wDUkQichF8PHNvwnFqsVmU5eCg8
NPj2K+dvfbOnOn6FzMWU6flrFQZQm/HCB0igqyd85kfb9/Cj9iT2Jf9hRA1fhhBE
q+VyHISSD9QOO3E3S8mrHMc6otZdIwT8MWMh0YaJsYSV4/cPe9nVODV9PLPkVgz4
COZcGYfBDElvWo5IPZRcD47rdJx9j2eqnE4dhAOsFSH2qgsQKU566ZAX14DUUCZ6
Pk+6uhRdAoGBAPCyoK94iZ5A9+wc9MdpgyRKlq7RA2zt0GR+47+Liv614+zJAnc1
qIdEqIcDPdMILQbWBIMwtwxYiSQ1ILuL6IfMIsXz3Jpd+BGnV3EtJFg0/TshPu/U
hQsgOjShajemzP8oT3AGcqWmCa6xH27wbPgT9gIa7XkS64kRqBOLZCATAoGBAOYK
O3J0RDgUTWWWBT2INNIn9S1/B7TmHbER3ESCP+f0k5GRvK5LLAqWNrrdGWN5s/xQ
tRWMYOfYDbkyAsolS3vKpo/m3upkTD/G8J0faA/P30wvq3XOEKstbqeiYlAofeVz
TI1sHAutJx0OtOdQB9L/YDbb7JCqY3dBZqeG2Bj3AoGAWyF2fxknGZjFhUrtTnbf
ZSUsaeHO5zYfGX2FYydFJ+zb7/GnElVpilVvbTbH+Jd23Mi5CqauF2mJ+wB2dSui
jY+3drU+x99eJejyzXHm+dKOMg4DUzBmcvDvuK+IlKt9n/m2Idb/H0J/FfoPyaQT
zdVY5jElyhpkvQ/HUCcNcKsCgYEAqUfQVTpP1UDDb8UFGDG6RQhYeOFo52sLYHk3
MUb0BhpJ2a54PX5d63B81+fKPhSaKUuu5iuGSDYKa0TtHppxQKhxB6YqViv2nFwm
RkmNR01+Ec9mimtYgs8NBdkOJdSWSJofNFbhEIqcJNrkru+Kwm9g+x29qPtp9KEx
DIjDTyMCgYEAkdv4S187YxFAXGgcHAKVBP1awardfyspc1sUBa/ABGQl6JRLlI/W
xNCvRhfRxa5mCCLYhCRXIm74JEQC7craX+K0ln83YZ7870HuwNEanuqkIRQhMP1g
qA2xRVxCNmkZ/Ju6i1dXwmVomoECSAYRLqHbRHScSfqqFr2rAec9BQg=
-----END RSA PRIVATE KEY-----"

let public_key = "-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA2EobPLULbXZNI9GBoXvX
2X4Yplh4KIhDBWd33ll4id+Wsw6/vIvOKy2rB74zcTgmRi6ZVNArgX5jdC6+nJHe
r33DQl/CNls24n7H5b/gC9O253roJjYQU9FiwMH4nsneAJimXizNc00E5pqhvMbC
u1/ZoKb6hLBrcLF1KA3n38GTcmaqr9yI5TM65B2pKelNAbBaIB2oBWAKjPfvDLkO
hyNrgAKV677LD3fsU+lC+ea9ylt/KPoicOAQFZer1kqYcI+IeouG2ePJcPEH3xpH
0mFtaTzmJd5Hzr84kmM2NZRN+oOlOYOKFluSTWQ1aojPQBmhceQFnfLAKoOW4qS6
VQIDAQAB
-----END PUBLIC KEY-----"

let () =
  let jwt = Jwt.create () in
  let now = Unix.time () |> int_of_float in
  Printf.printf "NOW: %d\n" now;
  Jwt.set_alg jwt RS256 ~key:private_key;
  Jwt.add_grant jwt "iss" "https://www.googleapis.com/robot/v1/metadata/x509/recommendation-tracking%40recommendation-tracking.iam.gserviceaccount.com";
  Jwt.add_grant jwt "scope" "https://www.googleapis.com/auth/prediction";
  Jwt.add_grant jwt "aud" "https://www.googleapis.com/oauth2/v4/token";
  Jwt.add_grant_int jwt "exp" (now + 3600);
  Jwt.add_grant_int jwt "iat" now;
  Printf.printf "%s\n" @@ Jwt.dump ~pretty:true jwt;
  Printf.printf "%s\n" (Jwt.get_grant jwt "iss" |> function Some s -> s | None -> "<None>");
  Printf.printf "%d\n" @@ Jwt.get_grant_int jwt "iat";
  let jwt_encoded = Jwt.encode jwt in
  Printf.printf "\n\nEncoded:\n%s\n" jwt_encoded;
  Printf.printf "DECODED JWT: %s\n" (Jwt.dump ~pretty:true @@ Jwt.decode jwt_encoded ~key:public_key);
  Gc.full_major ();
  Printf.printf "DONE!\n"

let create_google_auth_payload ~key ~client_id =
  let now = Unix.time () |> int_of_float in
  let jwt = Jwt.create () in
  Jwt.set_alg jwt RS256 ~key;
  Jwt.add_grant jwt "aud" "https://www.googleapis.com/oauth2/v4/token";
  Jwt.add_grant jwt "iss" client_id;
  Jwt.add_grant jwt "scope" "https://www.googleapis.com/auth/cloud-platform";
  Jwt.add_grant_int jwt "exp" (now + 36);
  Jwt.add_grant_int jwt "iat" now;
  Jwt.encode jwt

let authenticate_with_google ~private_key ~client_email =
  let assertion = create_google_auth_payload ~key:private_key ~client_id:client_email in
  let headers = Header.init_with "content-type" "application/x-www-form-urlencoded" in
  let body = Uri.encoded_of_query [
    ("assertion", [assertion]);
    ("grant_type", ["urn:ietf:params:oauth:grant-type:jwt-bearer"])
  ] |> Cohttp_lwt.Body.of_string in
  Client.post ~headers ~body (Uri.of_string "https://www.googleapis.com/oauth2/v4/token") >>= fun (rsp, rsp_body) ->
  let code = rsp |> Response.status |> Code.code_of_status in
  Printf.printf "Response code: %d\n" code;
  Printf.printf "Response Headers: %s\n" (rsp |> Response.headers |> Header.to_string);
  rsp_body |> Cohttp_lwt.Body.to_string >|= fun body ->
  Printf.printf "Body of length: %d\n" (String.length body);
  Printf.printf "Body: %s\n" body

(*
let () =
  Lwt_main.run (authenticate_with_google
    ~private_key:"-----BEGIN PRIVATE KEY-----\n[INSERT REAL KEY HERE]\n-----END PRIVATE KEY-----"
    ~client_email:"[INSERT REAL CLIENT EMAIL HERE")
  |> fun hest ->
      ignore hest
*)


src/jwt.ml
type t
exception Jwt_error of string * int

let () = Callback.register_exception "jwt exception" (Jwt_error ("<function name>", -1))

type algorithm =
  | NONE
  | HS256
  | HS384 
  | HS512 
  | RS256 
  | RS384 
  | RS512 
  | ES256 
  | ES384 
  | ES512 
  | TERM

external create: unit -> t = "ocaml_jwt_create"
external decode: ?key: string -> string -> t = "ocaml_jwt_decode"
external dump: ?pretty:bool -> t -> string = "ocaml_jwt_dump"
external add_grant: t -> string -> string -> unit = "ocaml_jwt_add_grant"
external add_grant_int: t -> string -> int -> unit = "ocaml_jwt_add_grant_int"
external add_grant_bool: t -> string -> bool -> unit = "ocaml_jwt_add_grant_bool"

external get_grant : t -> string -> string option = "ocaml_jwt_get_grant"
external get_grant_int : t -> string -> int = "ocaml_jwt_get_grant_int"
external get_grant_bool : t -> string -> bool = "ocaml_jwt_get_grant_bool"

external encode : t -> string = "ocaml_jwt_encode"

external set_alg : ?key: string -> t -> algorithm -> unit = "ocaml_jwt_set_alg"



src/jwt_c.ml
type t

type error = [ `Jwt_error of int ]

type algorithm =
  | NONE
  | HS256
  | HS384 
  | HS512 
  | RS256 
  | RS384 
  | RS512 
  | ES256 
  | ES384 
  | ES512 
  | TERM

external create : unit -> (t, [> error ]) result = "ocaml_jwt_create"
external encode : t -> (string, [> error ]) result = "ocaml_jwt_encode"
external decode : ?key:string -> string -> (t, [> error ]) result = "ocaml_jwt_decode"
external dump: ?pretty:bool -> t -> string = "ocaml_jwt_dump"

external add_grant : t -> string -> string -> (unit, [> error ]) result = "ocaml_jwt_add_grant"
external add_grant_int : t -> string -> int -> (unit, [> error ]) result = "ocaml_jwt_add_grant_int"
external add_grant_bool : t -> string -> bool -> (unit, [> error ]) result = "ocaml_jwt_add_grant_bool"
external add_grants_json : t -> string -> (unit, [> error ]) result = "ocaml_jwt_add_grants_json"

external get_grant : t -> string -> string option = "ocaml_jwt_get_grant"
external get_grant_int : t -> string -> int option = "ocaml_jwt_get_grant_int"
external get_grant_bool : t -> string -> bool option = "ocaml_jwt_get_grant_bool"
external get_grants_json : ?key:string -> t -> string option = "ocaml_jwt_get_grants_json"

external del_grant : t -> string -> (unit, [> error ]) result = "ocaml_jwt_del_grant"
external del_grants : t -> (unit, [> error ]) result = "ocaml_jwt_del_grants"

external set_alg : ?key:string -> t -> algorithm -> (unit, [> error ]) result = "ocaml_jwt_set_alg"
external get_alg : t -> algorithm = "ocaml_jwt_get_alg"


main.ml
open Lwt
open Cohttp
open Cohttp_lwt_unix

(* the key is just something I generated with `openssl genrsa -out private.pem 2048` *)
let private_key = "-----BEGIN RSA PRIVATE KEY-----
MIIEpQIBAAKCAQEA2EobPLULbXZNI9GBoXvX2X4Yplh4KIhDBWd33ll4id+Wsw6/
vIvOKy2rB74zcTgmRi6ZVNArgX5jdC6+nJHer33DQl/CNls24n7H5b/gC9O253ro
JjYQU9FiwMH4nsneAJimXizNc00E5pqhvMbCu1/ZoKb6hLBrcLF1KA3n38GTcmaq
r9yI5TM65B2pKelNAbBaIB2oBWAKjPfvDLkOhyNrgAKV677LD3fsU+lC+ea9ylt/
KPoicOAQFZer1kqYcI+IeouG2ePJcPEH3xpH0mFtaTzmJd5Hzr84kmM2NZRN+oOl
OYOKFluSTWQ1aojPQBmhceQFnfLAKoOW4qS6VQIDAQABAoIBAQCh8NJ14KP+wD2/
nMSvUNdrS6NPIRxOVL/BkxfHzjuXF2ZxCJD8rbyezaGpuXowwja+A3PgccCxQx1Z
xwoGlp0hzkrdLm9uXVs5uG0ZE1G/6TOgG4En4wDUkQichF8PHNvwnFqsVmU5eCg8
NPj2K+dvfbOnOn6FzMWU6flrFQZQm/HCB0igqyd85kfb9/Cj9iT2Jf9hRA1fhhBE
q+VyHISSD9QOO3E3S8mrHMc6otZdIwT8MWMh0YaJsYSV4/cPe9nVODV9PLPkVgz4
COZcGYfBDElvWo5IPZRcD47rdJx9j2eqnE4dhAOsFSH2qgsQKU566ZAX14DUUCZ6
Pk+6uhRdAoGBAPCyoK94iZ5A9+wc9MdpgyRKlq7RA2zt0GR+47+Liv614+zJAnc1
qIdEqIcDPdMILQbWBIMwtwxYiSQ1ILuL6IfMIsXz3Jpd+BGnV3EtJFg0/TshPu/U
hQsgOjShajemzP8oT3AGcqWmCa6xH27wbPgT9gIa7XkS64kRqBOLZCATAoGBAOYK
O3J0RDgUTWWWBT2INNIn9S1/B7TmHbER3ESCP+f0k5GRvK5LLAqWNrrdGWN5s/xQ
tRWMYOfYDbkyAsolS3vKpo/m3upkTD/G8J0faA/P30wvq3XOEKstbqeiYlAofeVz
TI1sHAutJx0OtOdQB9L/YDbb7JCqY3dBZqeG2Bj3AoGAWyF2fxknGZjFhUrtTnbf
ZSUsaeHO5zYfGX2FYydFJ+zb7/GnElVpilVvbTbH+Jd23Mi5CqauF2mJ+wB2dSui
jY+3drU+x99eJejyzXHm+dKOMg4DUzBmcvDvuK+IlKt9n/m2Idb/H0J/FfoPyaQT
zdVY5jElyhpkvQ/HUCcNcKsCgYEAqUfQVTpP1UDDb8UFGDG6RQhYeOFo52sLYHk3
MUb0BhpJ2a54PX5d63B81+fKPhSaKUuu5iuGSDYKa0TtHppxQKhxB6YqViv2nFwm
RkmNR01+Ec9mimtYgs8NBdkOJdSWSJofNFbhEIqcJNrkru+Kwm9g+x29qPtp9KEx
DIjDTyMCgYEAkdv4S187YxFAXGgcHAKVBP1awardfyspc1sUBa/ABGQl6JRLlI/W
xNCvRhfRxa5mCCLYhCRXIm74JEQC7craX+K0ln83YZ7870HuwNEanuqkIRQhMP1g
qA2xRVxCNmkZ/Ju6i1dXwmVomoECSAYRLqHbRHScSfqqFr2rAec9BQg=
-----END RSA PRIVATE KEY-----"

let public_key = "-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA2EobPLULbXZNI9GBoXvX
2X4Yplh4KIhDBWd33ll4id+Wsw6/vIvOKy2rB74zcTgmRi6ZVNArgX5jdC6+nJHe
r33DQl/CNls24n7H5b/gC9O253roJjYQU9FiwMH4nsneAJimXizNc00E5pqhvMbC
u1/ZoKb6hLBrcLF1KA3n38GTcmaqr9yI5TM65B2pKelNAbBaIB2oBWAKjPfvDLkO
hyNrgAKV677LD3fsU+lC+ea9ylt/KPoicOAQFZer1kqYcI+IeouG2ePJcPEH3xpH
0mFtaTzmJd5Hzr84kmM2NZRN+oOlOYOKFluSTWQ1aojPQBmhceQFnfLAKoOW4qS6
VQIDAQAB
-----END PUBLIC KEY-----"

let () =
  begin
    let jwt = Jwt.create () in
    let now = Unix.time () |> int_of_float in
    Printf.printf "NOW: %d\n" now;
    Jwt.set_alg jwt JWT_ALG_RS256 ~key:private_key;
    Jwt.add_grant jwt "iss" "https://www.googleapis.com/robot/v1/metadata/x509/recommendation-tracking%40recommendation-tracking.iam.gserviceaccount.com";
    Jwt.add_grant jwt "scope" "https://www.googleapis.com/auth/prediction";
    Jwt.add_grant jwt "aud" "https://www.googleapis.com/oauth2/v4/token";
    Jwt.add_grant_int jwt "exp" (now + 3600);
    Jwt.add_grant_int jwt "iat" now;
    Printf.printf "%s\n" @@ Jwt.dump ~pretty:true jwt;
    Printf.printf "%s\n" (Jwt.get_grant jwt "iss" |> function Some s -> s | None -> "<None>");
    Printf.printf "%d\n" @@ Jwt.get_grant_int jwt "iat";
    let jwt_encoded = Jwt.encode jwt in
    Printf.printf "\n\nEncoded:\n%s\n" jwt_encoded; 
    Printf.printf "DECODED JWT: %s\n" (Jwt.dump ~pretty:true @@ Jwt.decode jwt_encoded ~key:public_key)
  end;
  Gc.full_major ();
  Printf.printf "DONE!\n"

let create_google_auth_payload ~key ~client_id =
    let now = Unix.time () |> int_of_float in
    let jwt = Jwt.create () in
    Jwt.set_alg jwt JWT_ALG_RS256 ~key;
    Jwt.add_grant jwt "aud" "https://www.googleapis.com/oauth2/v4/token";
    Jwt.add_grant jwt "iss" client_id;
    Jwt.add_grant jwt "scope" "https://www.googleapis.com/auth/cloud-platform";
    Jwt.add_grant_int jwt "exp" (now + 36);
    Jwt.add_grant_int jwt "iat" now;
    Jwt.encode jwt

let authenticate_with_google ~private_key ~client_email =
  let assertion = create_google_auth_payload ~key:private_key ~client_id:client_email in
  let headers = Header.init_with "content-type" "application/x-www-form-urlencoded" in
  let body = Uri.encoded_of_query [
    ("assertion", [assertion]);
    ("grant_type", ["urn:ietf:params:oauth:grant-type:jwt-bearer"])
  ] |> Cohttp_lwt.Body.of_string in
  Client.post ~headers ~body (Uri.of_string "https://www.googleapis.com/oauth2/v4/token") >>= fun (rsp, rsp_body) ->
  let code = rsp |> Response.status |> Code.code_of_status in
  Printf.printf "Response code: %d\n" code;
  Printf.printf "Response Headers: %s\n" (rsp |> Response.headers |> Header.to_string);
  rsp_body |> Cohttp_lwt.Body.to_string >|= fun body ->
  Printf.printf "Body of length: %d\n" (String.length body);
  Printf.printf "Body: %s\n" body

(*
let () =
  Lwt_main.run (authenticate_with_google
    ~private_key:"-----BEGIN PRIVATE KEY-----\n[INSERT REAL KEY HERE]\n-----END PRIVATE KEY-----"
    ~client_email:"[INSERT REAL CLIENT EMAIL HERE")
  |> fun hest ->
      ignore hest
*)
